bundle/edu.gemini.itc.shared/src/main/java/edu/gemini/itc/shared/ConfigCreator.java

```diff
@@ -0,0 +1,201 @@
+    public static final ItemKey AOFieldLensKey = new ItemKey("adaptive optics:fieldLens");
+
+    private final ItcParameters itcParams;
+    private final ObservationDetails obsDetailParams;
+
+    public ConfigCreator(final ItcParameters p) {
+        this.itcParams = p;
+        this.obsDetailParams = p.observation();
+    }
+
+    // create part of config common for all instruments
+    public Config[] createCommonConfig(int numberExposures) {
+        final CalculationMethod calcMethod = obsDetailParams.calculationMethod();
+        int numberCoadds = calcMethod.coaddsOrElse(1);
+        Config[] dc = new DefaultConfig[numberExposures];
+        String offset = Double.toString(calcMethod.offset());
+        // for spectroscopic observations we consider ABBA offset pattern
+        List<String> spectroscopyOffsets = new ArrayList<>(Arrays.asList("0", offset, offset, "0"));
+        // for imaging observations we consider ABAB offset pattern
+        List<String> imagingOffsets = new ArrayList<>(Arrays.asList("0", offset, "0", offset));
+        List<String> offsetList = new ArrayList<>();
+
+        for (int i = 0; i < (1 + numberExposures / 4); i++) {
+            if (calcMethod instanceof Imaging) {
+                offsetList.addAll(imagingOffsets);
+            } else if (calcMethod instanceof Spectroscopy) {
+                offsetList.addAll(spectroscopyOffsets);
+            }
+        }
+
+        for (int i = 0; i < dc.length; i++) {
+            Config step = new DefaultConfig();
+            dc[i] = step;
+
+            step.putItem(ExposureTimeKey, obsDetailParams.exposureTime());
+            step.putItem(ObserveTypeKey, "OBJECT");
+            step.putItem(ObserveClassKey, "SCIENCE");
+            step.putItem(CoaddsKey, numberCoadds);
+            step.putItem(TelescopePKey, "0");
+            step.putItem(TelescopeQKey, offsetList.get(i));
+
+            if (itcParams.telescope().getWFS().equals(GuideProbe.Type.PWFS)) {
+                step.putItem(GuideWithPWFS2Key, "guide");
+            }
+        }
+        return dc;
+    }
+
+    /**
+     * Instrument-specific configs
+     */
+    public Config[] createGnirsConfig(GnirsParameters gnirsParams, int numExp) {
+        Config[] conf = createCommonConfig(numExp);
+
+        for (Config step : conf) {
+            step.putItem(ReadModeKey, (gnirsParams.readMode()));
+            step.putItem(InstInstrumentKey, "GNIRS");
+            step.putItem(SlitWidthKey, gnirsParams.slitWidth().logValue());
+
+            if (gnirsParams.altair().isDefined()) {
+                AltairParameters altairParameters = gnirsParams.altair().get();
+                step.putItem(AOGuideStarTypeKey, altairParameters.wfsMode().displayValue());
+                step.putItem(AOSystemKey, "Altair");
+                step.putItem(AOFieldLensKey, altairParameters.fieldLens().displayValue());
+            }
+        }
+        return conf;
```

Comment:
This could again be simplified using a `Config` with the repeated values in the `for` loop and then for each step, using `putAll`, e.g.:

```java
final Config commonStep = new DefaultConfig();
commonStep.putItem(ReadModeKey, (gnirsParams.readMode()));
commonStep.putItem(InstInstrumentKey, "GNIRS");
commonStep.putItem(SlitWidthKey, gnirsParams.slitWidth().logValue());
if (gnirsParams.altair().isDefined()) {
  AltairParameters altairParameters = gnirsParams.altair().get();   step.putItem(AOGuideStarTypeKey, altairParameters.wfsMode().displayValue());
  commonStep.putItem(AOSystemKey, "Altair");
  commonStep.putItem(AOFieldLensKey, altairParameters.fieldLens().displayValue());
}

final Config[] conf = createCommonConfig(numExp);
for (final Config step : conf) {
  step.putAll(commonStep);
}
```

Again, though, I'm not clear why you're putting the exact same step repeatedly into an array. That may be the proper behaviour but it seems a but unusual to me.

Reply:
Also, same for the below `createXXXXConfig` methods.

Reply:
Well, as you noted above, the steps are not "exactly the same", the offsets are different. For the GSAOI (still in progress) there will be additional differences. I did it this way in order to reuse the OT methods for overhead calculations with as little hacking as possible. After a lot of consideration I decided that this way (OT sequence-like) it would be easier to maintain and change, in case of extra requirements, or if the ITC gets more sophisticated.